### PR TITLE
Issue 349 documentation sufficiency checks

### LINKED ISSUE #349: SAR PL-2 Documentation Sufficiency Checks
**Extended Description**
- As a ... , in order to ... , I want ...

**Preconditions**
- Preconditions...

**Acceptance Criteria**
- [ ] a check that targets specifically when there are multiple findings for control PL-2 in the `risk`s
- [ ] a check that asserts there is not a large number of findings around insufficient documentation for PL-2

**Story Tasks**
- [ ] Tasks...

**Definition of Done**
 - [ ] Acceptance criteria met - Each user story should meet the acceptance criteria in the description
 - [ ] Unit test coverage of our code > 90%  (*from [QASP](https://github.com/flexion/TTS-10x/wiki/TTS-QASP)*) this may be fuzzy and hard to prove
 - [ ] Code quality checks passed (*from [QASP](https://github.com/flexion/TTS-10x/wiki/TTS-QASP)*)
 - [ ] Accessibility: (*from [QASP](https://github.com/flexion/TTS-10x/wiki/TTS-QASP)*) as we create guidance or documentation and reports (semantic tagging including aria tags): demonstrate with 0 errors reported for WCAG 2.1 AA standards using an automated scanner and 0 errors reported in manual testing
 - [ ] Code reviewed - Code reviewed by at least one other team members (or developed by a pair)
 - [ ] Source code merged - Code that’s demoed must be in source control and merged
 - [ ] Code must successfully build and deploy into staging environment (*from [QASP](https://github.com/flexion/TTS-10x/wiki/TTS-QASP)*): this may evolve from xslt sh pipline into something more
 - [ ] Security reviewed and reported - Conduct vulnerability and compliance scanning. threat modeling? 
 - [ ] Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities (*from [QASP](https://github.com/flexion/TTS-10x/wiki/TTS-QASP)*)
 - [ ] Usability tests passed - Each user story should be easy to use by target users (development community? FedRAMP FART team)
 - [ ] Usability testing and other user research methods must be conducted at regular intervals throughout the development process (not just at the beginning or end). (*from [QASP](https://github.com/flexion/TTS-10x/wiki/TTS-QASP)*)
 - [ ] Code refactored for clarity - Code must be clean, self-documenting
 - [ ] No local design debt
 - [ ] Load/performance tests passed - test data needed - saxon instrumentation
 - [ ] Documentation generated - update readme or contributing markdown as necessary.
 - [ ] Architectural Decision Record completed as necessary for significant design choices

### PR DESCRIPTION
**Pull request by markXLIX**
Closes #349 
Compare catalog objectives for PL-2 against SAR PL-2 objectives that are not satisfied

Ensure there are no more than one finding with a PL-2 target

### CODE REVIEW

File: src/validations/rules/sar.sch [hunk 1]
```diff
@@ -377,6 +377,34 @@
...
+                test="count($pl-2-other-than-satisfied-findings) gt 0"
+                unit:override-xspec="both">Every unsatisfied PL-2 objective must have a matching resolved baseline profile catalog part.</sch:assert>
+
+            <sch:assert
+                diagnostics="multiple-control-PL-2-findings-diagnostic"
+                fedramp:specific="true"
+                id="multiple-control-PL-2-findings"
+                role="error"
+                test="count(oscal:finding[oscal:target[@type = 'objective-id' and contains(@target-id, 'pl-2')]]) lt 2">The SAR contains no more than
+                one finding for the control PL-2.</sch:assert>
```
Comment (GaryGapinski) [inline src/validations/rules/sar.sch:406]: The `id` and `diagnostics` identifiers are the inverse of the desired outcome.
The assertion message is for the wrong context the context is `result`, not SAR.
_Unsatisfied_ findings are important, not just any finding.

```suggestion
            <sch:assert
                diagnostics="minimal-control-PL-2-findings-diagnostic"
                fedramp:specific="true"
                id="minimal-control-PL-2-findings"
                role="error"
                test="count(oscal:finding[oscal:target[@type = 'objective-id'and status/@state ne 'satisfied' and contains(@target-id, 'pl-2')]]) lt 2">Every result contains no more than
                one unsatisfied finding for the control PL-2.</sch:assert>
```

**Review by GaryGapinski — CHANGES_REQUESTED** [review]
Make sure that format and indent is used.
Make sure that assertion identifiers match the assertion's positive outcome.
Make sure that assertion text matches the assertion's positive outcome.
Strive to place assertions in a context as specific as possible.
All of the stuff (prior to the assertion `multiple-control-PL-2-findings"`) that references the"profile objectives" should be moved to a general (not just PL-2) context for `target` with assertions that ensure the targeted id exists in the resolved profile. That could be as simple as "the target-id is an id in the resolved profile".

File: src/validations/rules/sar.sch [hunk 2]
```diff
@@ -377,6 +377,34 @@
...
+                name="pl-2-not-matches-other-than-satisfied-findings"
+                value="$pl-2-other-than-satisfied-findings[not(. = $pl-2-profile-objectives)]" />
+
+            <sch:assert
+                diagnostics="multiple-control-PL-2-objectives-diagnostic"
+                fedramp:specific="true"
+                id="multiple-control-PL-2-objectives"
+                role="error"
+                test="count($pl-2-other-than-satisfied-findings) gt 0"
+                unit:override-xspec="both">Every unsatisfied PL-2 objective must have a matching resolved baseline profile catalog part.</sch:assert>
```
[the remark this answers is not in the document]
Comment (markXLIX) [reply src/validations/rules/sar.sch:398]: That wasn't the issue as stated.  If this is desired then a new issue should be created with appropriate reference to guidance stating that this is needed.

[the remark this answers is not in the document]
Comment (GaryGapinski) [reply src/validations/rules/sar.sch]: Neither is this. No matter - someone else can make the related improvements at a future time.

File: src/validations/rules/sar.sch [hunk 3]
```diff
@@ -379,7 +379,7 @@
 
             <sch:let
                 name="pl-2-other-than-satisfied-findings"
-                value="oscal:finding/oscal:target[@type = 'objective-id'][oscal:status ne 'satisfied'][contains(@target-id, 'pl-2')]/@target-id" />
+                value="oscal:finding/oscal:target[@type = 'objective-id'][oscal:status ne 'satisfied'][matches(@target-id, '^pl-2$')]/@target-id" />
```
Comment (GaryGapinski) [inline src/validations/rules/sar.sch:382]: Objective target identifiers are not the same as control identifiers. See, e.g., [this](https://github.com/18F/fedramp-automation/blob/master/dist/content/templates/sar/xml/FedRAMP-SAR-OSCAL-Template.xml#L1043).

The PL-2 objectives are
```xml
         <part id="pl-2_obj" name="objective">
            <p>Determine if the organization:</p>
            <part id="pl-2.a_obj" name="objective">
               <prop name="label" value="PL-2(a)"/>
               <p>develops a security plan for the information system that:</p>
               <part id="pl-2.a.1_obj" name="objective">
                  <prop ns="https://fedramp.gov/ns/oscal"
                        name="response-point"
                        value="You must fill in this response point."/>
                  <prop name="method" class="fedramp" value="EXAMINE"/>
                  <prop name="label" value="PL-2(a)(1)"/>
                  <p>is consistent with the organization’s enterprise architecture;</p>
               </part>
               <part id="pl-2.a.2_obj" name="objective">
                  <prop ns="https://fedramp.gov/ns/oscal"
                        name="response-point"
                        value="You must fill in this response point."/>
                  <prop name="method" class="fedramp" value="EXAMINE"/>
                  <prop name="label" value="PL-2(a)(2)"/>
                  <p>explicitly defines the authorization boundary for the system;</p>
               </part>
               <part id="pl-2.a.3_obj" name="objective">
                  <prop ns="https://fedramp.gov/ns/oscal"
                        name="response-point"
                        value="You must fill in this response point."/>
                  <prop name="method" class="fedramp" value="EXAMINE"/>
                  <prop name="label" value="PL-2(a)(3)"/>
                  <p>describes the operational context of the information system in terms of missions and business processes;</p>
               </part>
               <part id="pl-2.a.4_obj" name="objective">
                  <prop ns="https://fedramp.gov/ns/oscal"
                        name="response-point"
                        value="You must fill in this response point."/>
                  <prop name="method" class="fedramp" value="EXAMINE"/>
                  <prop name="label" value="PL-2(a)(4)"/>
                  <p>provides the security categorization of the information system including supporting rationale;</p>
               </part>
               <part id="pl-2.a.5_obj" name="objective">
                  <prop ns="https://fedramp.gov/ns/oscal"
                        name="response-point"
                        value="You must fill in this response point."/>
                  <prop name="method" class="fedramp" value="EXAMINE"/>
                  <prop name="label" value="PL-2(a)(5)"/>
                  <p>describes the operational environment for the information system and relationships with or connections to other information systems;</p>
               </part>
               <part id="pl-2.a.6_obj" name="objective">
                  <prop ns="https://fedramp.gov/ns/oscal"
                        name="response-point"
                        value="You must fill in this response point."/>
                  <prop name="method" class="fedramp" value="EXAMINE"/>
                  <prop name="label" value="PL-2(a)(6)"/>
                  <p>provides an overview of the security requirements for the system;</p>
               </part>
               <part id="pl-2.a.7_obj" name="objective">
                  <prop ns="https://fedramp.gov/ns/oscal"
                        name="response-point"
                        value="You must fill in this response point."/>
                  <prop name="method" class="fedramp" value="EXAMINE"/>
                  <prop name="label" value="PL-2(a)(7)"/>
                  <p>identifies any relevant overlays, if applicable;</p>
               </part>
               <part id="pl-2.a.8_obj" name="objective">
                  <prop ns="https://fedramp.gov/ns/oscal"
                        name="response-point"
                        value="You must fill in this response point."/>
                  <prop name="method" class="fedramp" value="EXAMINE"/>
                  <prop name="label" value="PL-2(a)(8)"/>
                  <p>describes the security controls in place or planned for meeting those requirements including a rationale for the tailoring and supplemental decisions;</p>
               </part>
               <part id="pl-2.a.9_obj" name="objective">
                  <prop ns="https://fedramp.gov/ns/oscal"
                        name="response-point"
                        value="You must fill in this response point."/>
                  <prop name="method" class="fedramp" value="INTERVIEW"/>
                  <prop name="method" class="fedramp" value="TEST"/>
                  <prop name="label" value="PL-2(a)(9)"/>
                  <p>is reviewed and approved by the authorizing official or designated representative prior to plan implementation;</p>
               </part>
            </part>
            <part id="pl-2.b_obj" name="objective">
               <prop name="label" value="PL-2(b)"/>
               <part id="pl-2.b_obj.1" name="objective">
                  <prop ns="https://fedramp.gov/ns/oscal"
                        name="response-point"
                        value="You must fill in this response point."/>
                  <prop name="method" class="fedramp" value="EXAMINE"/>
                  <prop name="label" value="PL-2(b)[1]"/>
                  <p>defines personnel or roles to whom copies of the security plan are to be distributed and subsequent changes to the plan are to be communicated;</p>
               </part>
               <part id="pl-2.b_obj.2" name="objective">
                  <prop ns="https://fedramp.gov/ns/oscal"
                        name="response-point"
                        value="You must fill in this response point."/>
                  <prop name="method" class="fedramp" value="INTERVIEW"/>
                  <prop name="method" class="fedramp" value="TEST"/>
                  <prop name="label" value="PL-2(b)[2]"/>
                  <p>distributes copies of the security plan and communicates subsequent changes to the plan to organization-defined personnel or roles;</p>
               </part>
            </part>
            <part id="pl-2.c_obj" name="objective">
               <prop name="label" value="PL-2(c)"/>
               <part id="pl-2.c_obj.1" name="objective">
                  <prop ns="https://fedramp.gov/ns/oscal"
                        name="response-point"
                        value="You must fill in this response point."/>
                  <prop name="method" class="fedramp" value="EXAMINE"/>
                  <prop name="label" value="PL-2(c)[1]"/>
                  <p>defines the frequency to review the security plan for the information system;</p>
               </part>
               <part id="pl-2.c_obj.2" name="objective">
                  <prop ns="https://fedramp.gov/ns/oscal"
                        name="response-point"
                        value="You must fill in this response point."/>
                  <prop name="method" class="fedramp" value="EXAMINE"/>
                  <prop name="label" value="PL-2(c)[2]"/>
                  <p>reviews the security plan for the information system with the organization-defined frequency;</p>
               </part>
            </part>
            <part id="pl-2.d_obj" name="objective">
               <prop ns="https://fedramp.gov/ns/oscal"
                     name="response-point"
                     value="You must fill in this response point."/>
               <prop name="method" class="fedramp" value="INTERVIEW"/>
               <prop name="method" class="fedramp" value="TEST"/>
               <prop name="label" value="PL-2(d)"/>
               <p>updates the plan to address:</p>
               <part id="pl-2.d_obj.1" name="objective">
                  <prop name="label" value="PL-2(d)[1]"/>
                  <p>changes to the information system/environment of operation;</p>
               </part>
               <part id="pl-2.d_obj.2" name="objective">
                  <prop name="label" value="PL-2(d)[2]"/>
                  <p>problems identified during plan implementation;</p>
               </part>
               <part id="pl-2.d_obj.3" name="objective">
                  <prop name="label" value="PL-2(d)[3]"/>
                  <p>problems identified during security control assessments;</p>
               </part>
            </part>
            <part id="pl-2.e_obj" name="objective">
               <prop ns="https://fedramp.gov/ns/oscal"
                     name="response-point"
                     value="You must fill in this response point."/>
               <prop name="method" class="fedramp" value="INTERVIEW"/>
               <prop name="method" class="fedramp" value="TEST"/>
               <prop name="label" value="PL-2(e)"/>
               <p>protects the security plan from unauthorized:</p>
               <part id="pl-2.e_obj.1" name="objective">
                  <prop name="label" value="PL-2(e)[1]"/>
                  <p>disclosure; and</p>
               </part>
               <part id="pl-2.e_obj.2" name="objective">
                  <prop name="label" value="PL-2(e)[2]"/>
                  <p>modification.</p>
               </part>
            </part>
         </part>
```

**Review by GaryGapinski — CHANGES_REQUESTED** [review]
There is a regex error at line 382.
The document has not been formatted and indented: there is superfluous whitespace at line 1534.

File: src/validations/rules/sar.sch [hunk 4]
```diff
@@ -382,6 +382,36 @@
                 test="oscal:attestation[oscal:part[@name = 'authorization-statements']/oscal:prop[@ns = 'https://fedramp.gov/ns/oscal' and @name = 'recommend-authorization']]">
                 There must exist an attestation with a part containing a property with a name of 'recommend-authorization'.</sch:assert>
 
+            <sch:let
+                name="pl-2-other-than-satisfied-findings"
+                value="oscal:finding/oscal:target[@type = 'objective-id'][oscal:status ne 'satisfied'][matches(@target-id, '^pl-2\.')]/@target-id" />
+            <sch:report test="true()">PL-2: <sch:value-of select="$pl-2-other-than-satisfied-findings"/></sch:report>
```
Comment (GaryGapinski) [inline src/validations/rules/sar.sch:388]: This looks like debugging code.

**Review by GaryGapinski — CHANGES_REQUESTED** [review]

File: src/validations/rules/sar.sch (hunk 4 above)
Comment (markXLIX) [reply src/validations/rules/sar.sch]: shoot.  Yes it is.  Thanks for the catch.

File: src/validations/rules/sar.sch [hunk 5]
```diff
@@ -382,6 +382,35 @@
...
+                test="count($pl-2-other-than-satisfied-findings) gt 0"
+                unit:override-xspec="both">Within a SAR, every unsatisfied PL-2 objective must have a matching resolved baseline profile catalog
+                part.</sch:assert>
+
+            <sch:assert
+                diagnostics="minimal-control-PL-2-findings-diagnostic"
+                fedramp:specific="true"
+                id="minimal-control-PL-2-findings"
+                role="error"
+                test="count(oscal:finding[oscal:target[@type = 'objective-id' and oscal:status/@state ne 'statisfied' and contains(@target-id, 'pl-2')]]) lt 2">Every
```
Comment (GaryGapinski) [inline src/validations/rules/sar.sch:411]: Unfortunate that the regex in line 387 was not employed here.

**Review by GaryGapinski — APPROVED** [review]

**Pull request merged**